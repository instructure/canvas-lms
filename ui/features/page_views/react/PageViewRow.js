/*
 * Copyright (C) 2021 - present Instructure, Inc.
 *
 * This file is part of Canvas.
 *
 * Canvas is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, version 3 of the License.
 *
 * Canvas is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
 * A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Affero General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>.
 */

import React from 'react'
import {TruncateText} from '@instructure/ui-truncate-text'
import {Text} from '@instructure/ui-text'
import I18n from 'i18n!page_views'
import FriendlyDatetime from '@canvas/datetime/react/components/FriendlyDatetime'
import {IconCheckMarkSolid} from '@instructure/ui-icons'
import {Table} from '@instructure/ui-table'
import UAParser from 'ua-parser-js'

export default class PageViewRow extends React.Component {
  static displayName = 'Row'

  parseUserAgentString(userAgent) {
    const SPEEDGRADER = 'SpeedGrader for iPad'
    const parser = new UAParser(userAgent)

    let browser = parser.getBrowser() ? parser.getBrowser().name : ''

    if (userAgent.toLowerCase().indexOf('speedgrader') >= 0) {
      browser = SPEEDGRADER
    }

    if (!browser) {
      browser = I18n.t('browsers.unrecognized', 'Unrecognized Browser')
    } else if (browser !== SPEEDGRADER) {
      const version = parser.getBrowser().version ? parser.getBrowser().version.split('.')[0] : '0'
      browser = `${browser} ${version}`
    }
    return browser
  }

  summarizedUserAgent(data) {
    return data.app_name || this.parseUserAgentString(data.user_agent)
  }

  readableInteractionTime(data) {
    const seconds = data.interaction_seconds
    if (seconds > 5) {
      return Math.round(seconds)
    } else {
      return '--'
    }
  }

  urlAsLink(data) {
    const method = data.http_method
    const url = data.url || ''

    if (!method || method !== 'get') return <>{url}</>
    else
      return (
        <>
          <a href={url}>{url}</a>
        </>
      )
  }

  particpatedIcon(data) {
    if (data.participated) return <IconCheckMarkSolid />
    else return <></>
  }

  renderRow() {
    return (
      <>
        <Table.Row key={this.props.rowData.session_id}>
          <Table.Cell key="url">
            <Text key="url_text">
              <TruncateText position="end">{this.urlAsLink(this.props.rowData)}</TruncateText>
            </Text>
          </Table.Cell>
          <Table.Cell key="created_at">
            <FriendlyDatetime
              key="datetime"
              dateTime={this.props.rowData.created_at}
              showTime
              format="%-d %b %Y %-l:%M%P"
            />
          </Table.Cell>
          <Table.Cell key="participated" textAlign="center">
            <Text>{this.particpatedIcon(this.props.rowData)}</Text>
          </Table.Cell>
          <Table.Cell key="interaction_seconds" textAlign="center">
            <Text>{this.readableInteractionTime(this.props.rowData)}</Text>
          </Table.Cell>
          <Table.Cell key="user_agent">
            <Text>{this.summarizedUserAgent(this.props.rowData)}</Text>
          </Table.Cell>
        </Table.Row>
      </>
    )
  }

  render() {
    if (this.props.rowData) return this.renderRow()
    else return <></>
  }
}
