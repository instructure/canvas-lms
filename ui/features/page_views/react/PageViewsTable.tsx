/*
 * Copyright (C) 2024 - present Instructure, Inc.
 *
 * This file is part of Canvas.
 *
 * Canvas is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, version 3 of the License.
 *
 * Canvas is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
 * A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Affero General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>.
 */

import React, {useRef} from 'react'
import {useScope as i18nScope} from '@canvas/i18n'
import useDateTimeFormat from '@canvas/use-date-time-format-hook'
import {Text} from '@instructure/ui-text'
import {Alert} from '@instructure/ui-alerts'
import {Table} from '@instructure/ui-table'
import {Spinner} from '@instructure/ui-spinner'
import {Tooltip} from '@instructure/ui-tooltip'
import doFetchApi from '@canvas/do-fetch-api-effect'
import {useInfiniteQuery, type QueryFunctionContext, type InfiniteData} from '@tanstack/react-query'
import {
  type APIPageView,
  type PageView,
  formatURL,
  formatInteractionTime,
  formatParticipated,
  formatUserAgent,
} from './utils'

export interface PageViewsTableProps {
  userId: string
  startDate?: Date
  endDate?: Date
}

type APIQueryParams = {
  page: string
  per_page: string
  start_time?: string
  end_time?: string
}

const I18n = i18nScope('page_views')

function UserAgentCell({view}: {view: PageView}): React.JSX.Element {
  return (
    <Tooltip renderTip={view.rawUserAgentString} on={['hover', 'focus']}>
      {view.userAgent}
    </Tooltip>
  )
}

export function PageViewsTable(props: PageViewsTableProps): React.JSX.Element {
  const observerRef = useRef<IntersectionObserver | null>(null)

  const formatDate = useDateTimeFormat('time.formats.short')
  async function fetchPageViews({
    pageParam = '1',
  }: QueryFunctionContext<[string, string, Date?], string>): Promise<{
    views: Array<PageView>
    nextPage: string | null
  }> {
    const params: APIQueryParams = {
      page: typeof pageParam === 'string' ? pageParam : '1',
      per_page: '50',
    }
    if (props.startDate) {
      if (!props.endDate) throw new RangeError('endDate must be set if startDate is set')
      params.start_time = props.startDate.toISOString()
      params.end_time = props.endDate.toISOString()
    }
    const path = `/api/v1/users/${props.userId}/page_views`
    const {json, link} = await doFetchApi<Array<APIPageView>>({path, params})
    if (typeof json === 'undefined') return {views: [], nextPage: null}
    const views: Array<PageView> = json.map(v => ({
      id: v.id,
      url: formatURL(v),
      createdAt: new Date(v.created_at),
      participated: formatParticipated(v),
      interactionSeconds: formatInteractionTime(v),
      rawUserAgentString: v.user_agent,
      userAgent: formatUserAgent(v),
    }))
    const nextPage = link?.next ? link.next.page : null
    return {views, nextPage}
  }

  function clearPageLoadTrigger() {
    if (observerRef.current === null) return
    observerRef.current.disconnect()
    observerRef.current = null
  }

  function setPageLoadTrigger(ref: Element | null) {
    if (ref === null) return
    clearPageLoadTrigger()
    observerRef.current = new IntersectionObserver(function (entries) {
      if (entries[0].isIntersecting) {
        fetchNextPage()
        clearPageLoadTrigger()
      }
    })
    observerRef.current.observe(ref)
  }

  const {data, fetchNextPage, isFetching, isFetchingNextPage, hasNextPage, isSuccess, error} =
    useInfiniteQuery<
      {views: Array<PageView>; nextPage: string | null},
      Error,
      InfiniteData<{views: Array<PageView>; nextPage: string | null}>,
      [string, string, Date?],
      string
    >({
      queryKey: ['page_views', props.userId, props.startDate],
      queryFn: fetchPageViews,
      staleTime: 10 * 60 * 1000, // 10 minutes
      getNextPageParam: lastPage => lastPage.nextPage,
      initialPageParam: '1',
    })

  if (isFetching && !isFetchingNextPage) return <Spinner renderTitle={I18n.t('Loading')} />

  if (!isSuccess) {
    // if we have an error, then the query failed, display an alert
    if (error) {
      let errorText: string
      const {response} = error as any // a response means an error from doFetchApi
      if (typeof response !== 'undefined')
        errorText = `API error: ${response.status} ${response.statusText}, fetching ${response.url}`
      else {
        const err = error as Error
        errorText = `Error formatting table: ${err.name}, ${err.message}`
      }
      return (
        <Alert variant="error" margin="small">
          <p>
            <strong>{I18n.t('Could not retrieve page views')}</strong>
            <br />
            {errorText}
          </p>
        </Alert>
      )
    }
    // if there is no error, then the query is still loading / retrying / something else
    // all we know is that an API fetch is not in progress but we still don't have data yet
    else return <Spinner renderTitle={I18n.t('Loading')} />
  }

  const uniqueViews: Record<string, PageView> = {}
  data.pages.forEach(page => {
    page.views.forEach(view => {
      uniqueViews[view.id] = view
    })
  })
  const viewKeys = Object.keys(uniqueViews)
  const isTriggerRow = (row: number) => row === viewKeys.length - 1 && !!hasNextPage && !isFetching
  const setTrigger = (row: number) =>
    isTriggerRow(row) ? (ref: Element | null) => setPageLoadTrigger(ref) : undefined

  return (
    <>
      <Table caption={I18n.t('Page views for this user')}>
        <Table.Head>
          <Table.Row>
            <Table.ColHeader id="page-view-url">{I18n.t('URL')}</Table.ColHeader>
            <Table.ColHeader id="page-view-date">{I18n.t('Date')}</Table.ColHeader>
            <Table.ColHeader id="page-view-participated" textAlign="center">
              {I18n.t('Participated')}
            </Table.ColHeader>
            <Table.ColHeader id="page-view-interaction-time" textAlign="end">
              {I18n.t('Time')}
            </Table.ColHeader>
            <Table.ColHeader id="page-view-user-agent">{I18n.t('User Agent')}</Table.ColHeader>
          </Table.Row>
        </Table.Head>
        <Table.Body data-testid="page-views-table-body">
          {viewKeys.map((id, idx) => {
            const v = uniqueViews[id]
            const url = (
              <Text size="small" elementRef={setTrigger(idx)}>
                {v.url}
              </Text>
            )
            return (
              <Table.Row data-testid="page-view-row" key={id}>
                <Table.Cell>{url}</Table.Cell>
                <Table.Cell>{formatDate(v.createdAt)}</Table.Cell>
                <Table.Cell textAlign="center">{v.participated}</Table.Cell>
                <Table.Cell textAlign="end">{v.interactionSeconds}</Table.Cell>
                <Table.Cell>
                  <UserAgentCell view={v} />
                </Table.Cell>
              </Table.Row>
            )
          })}
        </Table.Body>
      </Table>
      {isFetchingNextPage && <Spinner size="small" renderTitle={I18n.t('Loading')} />}
    </>
  )
}
