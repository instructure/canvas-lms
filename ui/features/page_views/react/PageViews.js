/*
 * Copyright (C) 2021 - present Instructure, Inc.
 *
 * This file is part of Canvas.
 *
 * Canvas is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, version 3 of the License.
 *
 * Canvas is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
 * A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Affero General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>.
 */

import React, {useState, useEffect, useCallback} from 'react'
import moment from 'moment'
import I18n from 'i18n!page_views'
import {Alert} from '@instructure/ui-alerts'
import {Text} from '@instructure/ui-text'
import {Link} from '@instructure/ui-link'
import {Flex} from '@instructure/ui-flex'
import {ScreenReaderContent} from '@instructure/ui-a11y-content'
import {Spinner} from '@instructure/ui-spinner'
import {Table} from '@instructure/ui-table'
import tz from '@canvas/timezone'
import CanvasDateInput from '@canvas/datetime/react/components/DateInput'
import {unfudgeDateForProfileTimezone} from '@canvas/datetime/date-functions'
import useFetchApi from '@canvas/use-fetch-api-hook'
import InfiniteScroll from '@canvas/infinite-scroll'
import PageViewRow from './PageViewRow'

const SIZE = 15

export const PageViews = ({userID}) => {
  const getCSVURL = useCallback(
    date => {
      const baseCSVURL = `/users/${userID}/page_views.csv`
      return date
        ? baseCSVURL +
            '?start_time=' +
            unfudgeDateForProfileTimezone(date).toISOString() +
            '&' +
            'end_time=' +
            moment(unfudgeDateForProfileTimezone(date)).add(1, 'days').toISOString()
        : baseCSVURL
    },
    [userID]
  )

  const [jsonData, setPageViewResults] = useState([])
  const [queryDate, updateQueryDate] = useState('')
  const [isLoading, setLoading] = useState(false)
  const [error, setError] = useState('')
  const [csvURLLink, setCSVURLLink] = useState(getCSVURL(''))
  const [page, setPage] = useState(1)
  const [hasMore, setHasMore] = useState(false)
  const [scrollContainer, setScrollContainer] = useState(null)

  const formatDate = date => {
    return tz.format(date, 'date.formats.medium')
  }

  useEffect(() => {
    if (error)
      <Alert
        liveRegion={() => document.getElementById('aria_alerts')}
        liveRegionPoliteness="assertive"
        screenReaderOnly
      >
        {error}
      </Alert>
  }, [error])

  useEffect(() => {
    setCSVURLLink(getCSVURL(queryDate))
  }, [getCSVURL, queryDate])

  const handlePageViewResults = useCallback(
    data => {
      setPageViewResults(currentJsonData => {
        setHasMore(data.length === SIZE)
        return currentJsonData.concat(data)
      })
    },
    [setPageViewResults]
  )

  useFetchApi(
    {
      success: handlePageViewResults,
      loading: setLoading,
      error: setError,
      path: `/api/v1/users/${userID}/page_views`,
      params: {
        start_time: `${queryDate ? unfudgeDateForProfileTimezone(queryDate).toISOString() : ''}`,
        end_time: `${
          queryDate
            ? moment(unfudgeDateForProfileTimezone(queryDate)).add(1, 'days').toISOString()
            : ''
        }`,
        per_page: SIZE,
        page
      },
      fetchAllPages: false
    },
    [queryDate]
  )

  const loadMore = () => {
    if (hasMore && !isLoading) {
      setPage(page + 1)
    }
  }

  const updateData = date => {
    setPageViewResults([])
    setPage(1)
    updateQueryDate(date)
  }

  return (
    <>
      <div style={{overflowx: 'hidden', overflowy: 'hidden'}}>
        <Flex
          key="pageviews_datefilter"
          direction="column"
          margin="none none small"
          align="start"
          overflowY="hidden"
          overflowX="hidden"
          height="570px"
        >
          <Flex.Item
            textAlign="start"
            padding="xx-small"
            height="50px"
            overflowY="hidden"
            overflowX="hidden"
          >
            <Text weight="bold">{I18n.t('Filter by date')}</Text>
          </Flex.Item>
          <Flex.Item
            padding="x-small"
            height="50px"
            shouldShrink
            overflowY="hidden"
            overflowX="hidden"
          >
            <Flex overflowY="hidden" overflowX="hidden">
              <Flex.Item>
                <CanvasDateInput
                  dataTestid="inputQueryDate"
                  renderLabel={
                    <ScreenReaderContent>{I18n.t('Filter by date')}</ScreenReaderContent>
                  }
                  onSelectedDateChange={updateData}
                  formatDate={formatDate}
                  withRunningValue
                />
              </Flex.Item>
              <Flex.Item padding="x-small">
                <Link id="page_views_csv_link" data-testid="page_views_csv_link" href={csvURLLink}>
                  {I18n.t('Download Page Views CSV')}
                </Link>
              </Flex.Item>
            </Flex>
          </Flex.Item>
          <Flex.Item height="20px" />
          <Flex.Item
            key="page_views_table"
            padding="x-small"
            height="450px"
            overflowY="auto"
            id="scrollContainer"
            data-testid="scrollContainer"
            elementRef={setScrollContainer}
          >
            <InfiniteScroll
              pageStart={0}
              loadMore={loadMore}
              hasMore={hasMore}
              scrollContainer={scrollContainer}
              loader={
                <Spinner
                  id="paginatedView-loading"
                  renderTitle={I18n.t('Loading')}
                  size="small"
                  margin="0 0 0 medium"
                />
              }
            >
              <Table layout="fixed" caption="Page View Results" user-id={userID}>
                <Table.Head>
                  <Table.Row key="0">
                    <Table.ColHeader id="URL">
                      {I18n.t('#page_views.table.headers.url', 'URL')}
                    </Table.ColHeader>
                    <Table.ColHeader id="Date" width="175px">
                      {(I18n.t('#page_views.table.headers.date'), 'Date')}
                    </Table.ColHeader>
                    <Table.ColHeader id="Paticipated" width="115px" textAlign="start">
                      {(I18n.t('#page_views.table.headers.participated'), 'Participated')}
                    </Table.ColHeader>
                    <Table.ColHeader id="Time" width="60px" textAlign="start">
                      {(I18n.t('#page_views.table.headers.time'), 'Time')}
                    </Table.ColHeader>
                    <Table.ColHeader id="Agent" width="125px">
                      {(I18n.t('#page_views.table.headers.user_agent'), 'User Agent')}
                    </Table.ColHeader>
                  </Table.Row>
                </Table.Head>
                <Table.Body id="page_view_results">
                  {jsonData.map((element, index) => (
                    <PageViewRow displayName="Row" rowData={element} key={index.toString()} />
                  ))}
                </Table.Body>
              </Table>
            </InfiniteScroll>
          </Flex.Item>
        </Flex>
      </div>
    </>
  )
}
