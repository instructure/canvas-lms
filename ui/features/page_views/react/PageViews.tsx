/*
 * Copyright (C) 2024 - present Instructure, Inc.
 *
 * This file is part of Canvas.
 *
 * Canvas is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, version 3 of the License.
 *
 * Canvas is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
 * A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Affero General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>.
 */

import React, {useState} from 'react'
import {useScope as i18nScope} from '@canvas/i18n'
import useDateTimeFormat from '@canvas/use-date-time-format-hook'
import {unfudgeDateForProfileTimezone} from '@instructure/moment-utils'
import CanvasDateInput2 from '@canvas/datetime/react/components/DateInput2'
import {PageViewsTable} from './PageViewsTable'
import {Flex} from '@instructure/ui-flex'
import {Button} from '@instructure/ui-buttons'
import {IconMsExcelLine} from '@instructure/ui-icons'

const I18n = i18nScope('page_views')
const icon = <IconMsExcelLine />

export interface PageViewsProps {
  userId: string
}

type DateRange = {
  date?: Date
  start?: Date
  end?: Date
}

export default function PageViews({userId}: PageViewsProps): React.JSX.Element {
  const [filterDate, setFilterDate] = useState<DateRange>({})
  const formatDateForDisplay = useDateTimeFormat('date.formats.long')
  const baseURL = `/users/${userId}/page_views.csv`

  function handleDateChange(date: Date | null) {
    if (date === null) {
      setFilterDate({})
      return
    }
    const start = unfudgeDateForProfileTimezone(date)
    if (start !== null) {
      const end = new Date(start.getTime() + 24 * 60 * 60 * 1000)
      setFilterDate({date, start, end})
    }
  }

  function downloadURL() {
    if (!filterDate.start || !filterDate.end) return baseURL
    return `${baseURL}?start_time=${filterDate.start.toISOString()}&end_time=${filterDate.end.toISOString()}`
  }

  return (
    <Flex direction="column">
      <Flex.Item padding="small">
        <CanvasDateInput2
          placeholder={I18n.t('Limit to a specific date')}
          selectedDate={filterDate.date?.toISOString()}
          formatDate={formatDateForDisplay}
          renderLabel={I18n.t('Filter by date')}
          onSelectedDateChange={handleDateChange}
          withRunningValue={true}
          interaction="enabled"
          dataTestid="page-views-date-filter"
        />
      </Flex.Item>
      <Flex.Item padding="small">
        <Button
          data-testid="page-views-csv-link"
          size="small"
          renderIcon={icon}
          href={downloadURL()}
        >
          {I18n.t('Download as CSV')}
        </Button>
      </Flex.Item>
      <Flex.Item>
        <PageViewsTable userId={userId} startDate={filterDate.start} endDate={filterDate.end} />
      </Flex.Item>
    </Flex>
  )
}
