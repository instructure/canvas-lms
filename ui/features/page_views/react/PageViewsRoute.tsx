/*
 * Copyright (C) 2024 - present Instructure, Inc.
 *
 * This file is part of Canvas.
 *
 * Canvas is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, version 3 of the License.
 *
 * Canvas is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
 * A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Affero General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>.
 */

import React from 'react'
import {useParams} from 'react-router-dom'
import {Portal} from '@instructure/ui-portal'
import PageViews from './PageViews'

export function Component(): React.JSX.Element | null {
  const {userId} = useParams<{userId: string}>()
  const mountPoint = document.getElementById('page_views_table')
  if (mountPoint === null || typeof userId === 'undefined') return null

  return (
    <Portal open={true} mountNode={mountPoint}>
      <PageViews userId={userId} />
    </Portal>
  )
}
