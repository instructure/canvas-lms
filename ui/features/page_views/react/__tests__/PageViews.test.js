/*
 * Copyright (C) 2021 - present Instructure, Inc.
 *
 * This file is part of Canvas.
 *
 * Canvas is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, version 3 of the License.
 *
 * Canvas is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
 * A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Affero General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>.
 */

import React from 'react'
import {shallow} from 'enzyme'
import {render, fireEvent} from '@testing-library/react'
import useFetchApi from '@canvas/use-fetch-api-hook'
import fetchMock from 'fetch-mock'
import {PageViews} from '../PageViews'

jest.mock('@canvas/use-fetch-api-hook')

const defaultProps = (props = {}) => ({
  userID: '718',
  ...props
})

afterEach(() => {
  fetchMock.restore()
})

it('renders correctly', () => {
  const wrapper = shallow(<PageViews {...defaultProps()} />)
  expect(wrapper).toMatchSnapshot()
})

it('Call api with error', () => {
  useFetchApi.mockImplementationOnce(({loading, error}) => {
    loading(false)
    error('Something Happened..')
  })
  const expectedDate = 'Jan 27, 2022'
  const {getByTestId} = render(<PageViews {...defaultProps()} />)
  const inputDate = getByTestId('inputQueryDate')
  fireEvent.change(inputDate, {target: {value: expectedDate}})
  fireEvent.blur(inputDate)
  expect(inputDate.value).toBe(expectedDate)
})

it('Show table without data', () => {
  useFetchApi.mockImplementationOnce(({loading, success}) => {
    loading(false)
    success([])
  })
  const {queryAllByRole} = render(<PageViews {...defaultProps()} />)
  expect(queryAllByRole('row')).toHaveLength(1)
})

it('Show data and load more data with scroll', () => {
  const defaultData = {
    action: 'new',
    app_name: null,
    asset_type: null,
    asset_user_access_id: null,
    context_type: 'Course',
    contributed: false,
    controller: 'discussion_topics',
    created_at: '2022-01-26T17:48:34Z',
    developer_key_id: null,
    http_method: 'get',
    id: '32ea0daf-70ac-434e-9111-cf7c71d23df0',
    interaction_seconds: 6,
    links: {user: 893, context: 11, asset: null, real_user: 1, account: 2},
    participated: true,
    remote_ip: '::1',
    render_time: 0.498169,
    session_id: 'd268f88b12896612544921c836302c8e',
    summarized: null,
    updated_at: '2022-01-26T17:48:34Z',
    url: null,
    user_agent:
      'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/97.0.4692.99 Safari/537.36',
    user_request: true
  }
  const data = Array(45)
    .fill()
    .map(index => {
      const copy = {...defaultData}
      copy.url = `url ${index}`
      return copy
    })
  useFetchApi.mockImplementationOnce(({loading, success}) => {
    loading(false)
    success(data)
  })
  const {getByTestId, queryAllByRole} = render(<PageViews {...defaultProps()} />)
  const scrollContainer = getByTestId('scrollContainer')
  scrollContainer.scrollTop = scrollContainer.scrollHeight
  expect(queryAllByRole('row')).toHaveLength(46)
})

it('Select the input date and reset the input date', () => {
  const data = [
    {
      action: 'new',
      app_name: null,
      asset_type: null,
      asset_user_access_id: null,
      context_type: 'Course',
      contributed: false,
      controller: 'discussion_topics',
      created_at: '2022-01-26T17:48:34Z',
      developer_key_id: null,
      http_method: 'get',
      id: '32ea0daf-70ac-434e-9111-cf7c71d23df0',
      interaction_seconds: 6,
      links: {user: 893, context: 11, asset: null, real_user: 1, account: 2},
      participated: true,
      remote_ip: '::1',
      render_time: 0.498169,
      session_id: 'd268f88b12896612544921c836302c8e',
      summarized: null,
      updated_at: '2022-01-26T17:48:34Z',
      url: null,
      user_agent:
        'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/97.0.4692.99 Safari/537.36',
      user_request: true
    },
    {
      action: 'new',
      app_name: null,
      asset_type: null,
      asset_user_access_id: null,
      context_type: 'Course',
      contributed: false,
      controller: 'discussion_topics',
      created_at: '2022-01-26T17:48:34Z',
      developer_key_id: null,
      http_method: 'post',
      id: '32ea0daf-70ac-434e-9111-cf7c71d23df0',
      interaction_seconds: 3,
      links: {user: 893, context: 11, asset: null, real_user: 1, account: 2},
      participated: null,
      remote_ip: '::1',
      render_time: 0.498169,
      session_id: 'd268f88b12896612544921c836302c8e',
      summarized: null,
      updated_at: '2022-01-26T17:48:34Z',
      url: 'http://localhost:3000/courses/11/discussion_topics/new',
      user_agent: 'Custom',
      user_request: true
    }
  ]
  useFetchApi.mockImplementationOnce(({loading, success}) => {
    loading(false)
    success(data)
  })
  const expectedDate = 'Jan 27, 2021'
  const {getByTestId} = render(<PageViews {...defaultProps()} />)
  const inputDate = getByTestId('inputQueryDate')
  fireEvent.change(inputDate, {target: {value: expectedDate}})
  fireEvent.blur(inputDate)
  expect(inputDate.value).toBe(expectedDate)
  fireEvent.change(inputDate, {target: {value: ''}})
  fireEvent.blur(inputDate)
  expect(inputDate.value).toBe('')
})

it('Select date and download file', () => {
  const data = [
    {
      action: 'new',
      app_name: null,
      asset_type: null,
      asset_user_access_id: null,
      context_type: 'Course',
      contributed: false,
      controller: 'discussion_topics',
      created_at: '2022-01-26T17:48:34Z',
      developer_key_id: null,
      http_method: 'get',
      id: '32ea0daf-70ac-434e-9111-cf7c71d23df0',
      interaction_seconds: 6,
      links: {user: 893, context: 11, asset: null, real_user: 1, account: 2},
      participated: true,
      remote_ip: '::1',
      render_time: 0.498169,
      session_id: 'd268f88b12896612544921c836302c8e',
      summarized: null,
      updated_at: '2022-01-26T17:48:34Z',
      url: null,
      user_agent:
        'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/97.0.4692.99 Safari/537.36',
      user_request: true
    }
  ]
  useFetchApi.mockImplementationOnce(({loading, success}) => {
    loading(false)
    success(data)
  })
  const expectedDate = 'Jan 27, 2022'
  const {getByTestId} = render(<PageViews {...defaultProps()} />)
  const inputDate = getByTestId('inputQueryDate')
  fireEvent.change(inputDate, {target: {value: expectedDate}})
  fireEvent.blur(inputDate)
  expect(inputDate.value).toBe(expectedDate)
  const pageViewLink = getByTestId('page_views_csv_link')
  fireEvent.click(pageViewLink)
  expect(pageViewLink).toBeInTheDocument()
})

it('Date with not result', () => {
  useFetchApi.mockImplementationOnce(({loading, success}) => {
    loading(false)
    success([])
  })
  const expectedDate = 'Jan 27, 2022'
  const {getByTestId, queryAllByRole} = render(<PageViews {...defaultProps()} />)
  const inputDate = getByTestId('inputQueryDate')
  fireEvent.change(inputDate, {target: {value: expectedDate}})
  fireEvent.blur(inputDate)
  expect(inputDate.value).toBe(expectedDate)
  expect(queryAllByRole('row')).toHaveLength(1)
})
