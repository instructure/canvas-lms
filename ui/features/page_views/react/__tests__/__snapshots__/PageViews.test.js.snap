// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<Fragment>
  <div
    style={
      Object {
        "overflowx": "hidden",
        "overflowy": "hidden",
      }
    }
  >
    <Flex
      align="start"
      as="span"
      direction="column"
      display="flex"
      elementRef={[Function]}
      height="570px"
      justifyItems="start"
      key="pageviews_datefilter"
      margin="none none small"
      overflowX="hidden"
      overflowY="hidden"
      withVisualDebug={false}
      wrap="no-wrap"
    >
      <Item
        as="span"
        elementRef={[Function]}
        height="50px"
        overflowX="hidden"
        overflowY="hidden"
        padding="xx-small"
        shouldGrow={false}
        shouldShrink={false}
        textAlign="start"
      >
        <Text
          as="span"
          letterSpacing="normal"
          size="medium"
          weight="bold"
          wrap="normal"
        >
          Filter by date
        </Text>
      </Item>
      <Item
        as="span"
        elementRef={[Function]}
        height="50px"
        overflowX="hidden"
        overflowY="hidden"
        padding="x-small"
        shouldGrow={false}
        shouldShrink={true}
      >
        <Flex
          as="span"
          direction="row"
          display="flex"
          elementRef={[Function]}
          justifyItems="start"
          overflowX="hidden"
          overflowY="hidden"
          withVisualDebug={false}
          wrap="no-wrap"
        >
          <Item
            as="span"
            elementRef={[Function]}
            shouldGrow={false}
            shouldShrink={false}
          >
            <CanvasDateInput
              dataTestid="inputQueryDate"
              formatDate={[Function]}
              onSelectedDateChange={[Function]}
              renderLabel={
                <ScreenReaderContent
                  as="span"
                >
                  Filter by date
                </ScreenReaderContent>
              }
              withRunningValue={true}
            />
          </Item>
          <Item
            as="span"
            elementRef={[Function]}
            padding="x-small"
            shouldGrow={false}
            shouldShrink={false}
          >
            <Link
              color="link"
              data-testid="page_views_csv_link"
              href="/users/718/page_views.csv"
              iconPlacement="start"
              id="page_views_csv_link"
              isWithinText={true}
            >
              Download Page Views CSV
            </Link>
          </Item>
        </Flex>
      </Item>
      <Item
        as="span"
        elementRef={[Function]}
        height="20px"
        shouldGrow={false}
        shouldShrink={false}
      />
      <Item
        as="span"
        data-testid="scrollContainer"
        elementRef={[Function]}
        height="450px"
        id="scrollContainer"
        key="page_views_table"
        overflowY="auto"
        padding="x-small"
        shouldGrow={false}
        shouldShrink={false}
      >
        <InfiniteScroll
          hasMore={false}
          loadMore={[Function]}
          loader={
            <Spinner
              as="div"
              id="paginatedView-loading"
              margin="0 0 0 medium"
              renderTitle="Loading"
              size="small"
              variant="default"
            />
          }
          pageStart={0}
          scrollContainer={null}
          threshold={250}
        >
          <Table
            caption="Page View Results"
            hover={false}
            layout="fixed"
            user-id="718"
          >
            <Head>
              <Row
                key="0"
              >
                <ColHeader
                  id="URL"
                  scope="col"
                  sortDirection="none"
                  textAlign="start"
                >
                  URL
                </ColHeader>
                <ColHeader
                  id="Date"
                  scope="col"
                  sortDirection="none"
                  textAlign="start"
                  width="175px"
                >
                  Date
                </ColHeader>
                <ColHeader
                  id="Paticipated"
                  scope="col"
                  sortDirection="none"
                  textAlign="start"
                  width="115px"
                >
                  Participated
                </ColHeader>
                <ColHeader
                  id="Time"
                  scope="col"
                  sortDirection="none"
                  textAlign="start"
                  width="60px"
                >
                  Time
                </ColHeader>
                <ColHeader
                  id="Agent"
                  scope="col"
                  sortDirection="none"
                  textAlign="start"
                  width="125px"
                >
                  User Agent
                </ColHeader>
              </Row>
            </Head>
            <Body
              id="page_view_results"
            />
          </Table>
        </InfiniteScroll>
      </Item>
    </Flex>
  </div>
</Fragment>
`;
